lib/src/lib.rs

```diff
@@ -1007,171 +459,164 @@ pub struct ColourCmd {
     /// LDraw commands parsed from the raw text content of the file.
     pub cmds: Vec<Command>,
 }
 
 /// Collection of [`SourceFile`] accessible from their reference filename.
 #[derive(Debug)]
 pub struct SourceMap {
-    /// Array of source files in the collection.
-    source_files: Vec<SourceFile>,
-
-    /// Map table of source files indices into [`SourceMap::source_files`] from
-    /// their reference filename, as it appears in [`SubFileRefCmd`].
-    filename_map: HashMap<String, usize>,
-}
-
-/// Reference to a single [`SourceFile`] instance in a given [`SourceMap`].
-#[derive(Debug, Copy, Clone, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
-pub struct SourceFileRef {
-    index: usize,
-    // TODO: ref to SourceMap?
-}
-
-impl CommentCmd {
-    pub fn new(text: &str) -> CommentCmd {
-        CommentCmd {
-            text: text.to_string(),
-        }
-    }
-}
-
-impl SourceFileRef {
-    pub fn get<'a>(&'a self, source_map: &'a SourceMap) -> &'a SourceFile {
-        source_map.get(*self)
-    }
+    /// Map of filenames to source files.
+    // TODO: How to handle case sensitivity like STUD.DAT vs stud.dat?
+    source_files: HashMap<String, SourceFile>,
 }
 
 impl SourceMap {
     /// Construct a new empty source map.
     pub fn new() -> SourceMap {
         SourceMap {
-            source_files: vec![],
-            filename_map: HashMap::new(),
+            source_files: HashMap::new(),
         }
     }
 
-    /// Get a reference to the source file corresponding to a [`SourceFileRef`].
-    fn get(&self, source_file_ref: SourceFileRef) -> &SourceFile {
-        &self.source_files[source_file_ref.index]
+    /// Returns a reference to the source file corresponding to `filename`.
+    pub fn get(&self, filename: &str) -> Option<&SourceFile> {
+        self.source_files.get(filename)
     }
 
-    /// Get a mutable reference to the source file corresponding to a [`SourceFileRef`].
-    fn get_mut(&mut self, source_file_ref: SourceFileRef) -> &mut SourceFile {
-        &mut self.source_files[source_file_ref.index]
+    /// Returns a mutable reference to the source file corresponding to `filename`.
+    pub fn get_mut(&mut self, filename: &str) -> Option<&mut SourceFile> {
+        self.source_files.get_mut(filename)
     }
 
-    /// Find a source file by its reference filename.
-    fn find_filename(&self, filename: &str) -> Option<SourceFileRef> {
-        self.filename_map
-            .get(filename)
-            .map(|&index| SourceFileRef { index })
-    }
-
-    /// Insert a new source file into the collection.
-    fn insert(&mut self, source_file: SourceFile) -> SourceFileRef {
-        if let Some(&index) = self.filename_map.get(&source_file.filename) {
-            SourceFileRef { index }
+    /// Inserts a new source file into the collection.
+    /// Returns a copy of `source_file` or the main file for multi-part documents (MPD).
+    pub fn insert(&mut self, source_file: SourceFile) -> SourceFile {
```

Comment:
That cloning thing again is quite ugly and inefficient. I'd rather return a reference to the file into the map, which can be either a simple `String` for now, or a newtype which normalized the key of the `HashMap` as discussed in another comment on casing. Then if the user is really interested in retrieving the source file (that they just inserted!) they can use `get()` or `get_mut()`, but there's a good chance the code inserting doesn't need to query immediately.

Reply:
Changed to just return `String`